modules/modelSetup/mixin/ModelSetupDiffusionLossMixin.py

```diff
@@ -149,28 +150,75 @@ def __unmasked_losses(
             losses /= mask_mean
 
         return losses
-
-    def __min_snr_weight(
+    
+    def calculate_loss(
```

Comment:
This is a bug. It will not work as you expect with multiple inheritance. Consider:

```py
 import abc
 class AbstractParent:
   @abc.abstractmethod
   def display(self):
     pass
 
 class RegularParent:
   def display(self):
     print("RegularParent")
 
 class Child(AbstractParent, RegularParent):
   pass

 x = Child()
 x.display()
```

There will be no output. Python method overriding searches left-to-right in the parent class list. Even if you went to the various children and made ModelSetupDiffusionLossMixin appear before BaseModelSetup, this would still be uncomfortably fragile.

You should instead create a subclass of BaseModelSetup, called BaseDiffusionModelSetup, that inherits BaseModelSetup and ModelSetupDiffusionLossMixin, and define calculate_loss in that child class. Then the other Base*Setups can inherit that class.

Reply:
I still think the best solution is to move the function call back into each (Base-)setup class. Since it's only a single function call, there is not much code duplication, but the intention is a lot clearer, with no added abstraction.

Reply:
I've gone back to the old code in https://github.com/Nerogar/OneTrainer/pull/212/commits/94506c6692da3eabdbba082e902b36747192feec.